[lldb][AArch64] Add isAArch64SMEFA64 check to SME testing

DavidSpickett opened the pull request:
FEAT_SME_FA64 (smefa64 in Linux cpuinfo) allows the use of the full A64 instruction set while in streaming SVE mode.

See https://developer.arm.com/documentation/ddi0616/latest/ for details.

This means for example if we want to write to the ffr register during or use floating point registers while in streaming mode, we need this extension.

I initially was using QEMU which has it by default, and switched to Arm's FVP which does not. So this change adds a more strict check and converts most of the tests to use that. It would be possible in some cases to avoid the offending instructions but it would be a lot of effort and liable to fail randomly as the C library changes.

It is also my assumption that the majority of systems will have smefa64 as QEMU has chosen to have. If I turn out to be wrong, we can make the effort to get the tests working without smefa64.

`isAArch64SME` remains for some tests, which are as follows:
* `test_aarch64_dynamic_regset_config` merely checks for the presence of a register set, which appears for any SME system not just one with smefa64.
* `test_aarch64_dynamic_regset_config_sme_za_disabled` only needs the ZA register and does not enter streaming mode.
* `test_sme_not_present` tests for the absence of the SME register set, so must be skipped if any form of SME is present.
* Various tests in `TestSVERegisters.py` need to know if SME is present at all to generate an expected SVCR value. Earlier in the callstack something else checked `isAArch64SMEFA64` already.
* `TestAArch64LinuxTLSRegisters.py` needs to test the `tpidr2` register if any form of SME is present. msr/mrs instructions are used to do this and are allowed even if smefa64 is not present.

Commit:
[lldb][AArch64] Add isAArch64SMEFA64 check to SME testing

FEAT_SME_FA64 (smefa64 in Linux cpuinfo) allows the use of the full
A64 instruction set while in streaming SVE mode.

See https://developer.arm.com/documentation/ddi0616/latest/ for
details.

This means for example if we want to write to the ffr register during
or use floating point registers while in streaming mode, we need this extension.

I initially was using QEMU which has it by default, and switched to
Arm's FVP which does not. So this change adds a more strict check and
converts most of the tests to use that.

It would be possible in some cases to avoid the offending instructions
but it would be a lot of effort and liable to fail randomly as the C
library changes.

It is also my assumption that the majority of systems will have
smefa64 as QEMU has chosen to have. If I turn out to be wrong, we
can make the effort to get the tests working without smefa64.

Commit:
Fix python formatting.

Commit:
Only read cpuinfo once.

omjavaid reviewed (commented):
After this change isAArch64SME check will be replaced on all locations? 

So the manual says "The SVE FFR predicate register is not architecturally visible when the PE is in Streaming SVE mode if FEAT_SME_FA64 is not implemented or not enabled at the current Exception level."

All SVE tests seem to use FFR does it mean isAArch64SME is now redundant as far as LLDB testing is concerned?

DavidSpickett commented:
I ran all the tests that originally had `isAArch64SME` in, on an FVP without smefa64 and added checks as needed. But I did not make note of the unchanged ones at the time, so I've redone that and added it to the commit message.

The ffr use you specifically noticed, the tests use isAArch64SMEFA64 up front and skip if it's not there. Then later when they want an SVCR value they just use isAArch64SME since strictly that's all that bit needs to know.

We could just have isAArch64SMEFA64 everywhere even if we didn't really need smefa64, but I figured it would be nice to have as many tests run in either situation as we can (for my benefit more than anything else right now).
